File: Python/boj/Gold/2470.py
import sys

if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    N = int(input())
    numbers = sorted(list(map(int, input().split(" "))))
    start = 0
    end = N - 1
    min = abs(numbers[start] + numbers[end])
    result = [numbers[start], numbers[end]]

    while start < end:
        sum = abs(numbers[start] + numbers[end])

        if sum < min:
            min = sum
            result = [numbers[start], numbers[end]]

            if sum == 0:
                break
        elif numbers[start] + numbers[end] > 0:
            end -= 1
        else:
            start += 1

    print(result[0], result[1])

File: JavaScript/boj/Gold/2470.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const N = +input[0];
const numbers = input[1].split(' ').map(Number);
let result = [];
let leftIdx = 0;
let rightIdx = N - 1;
let min = Infinity;

numbers.sort((a, b) => a - b);

while (leftIdx < rightIdx) {
  const sum = numbers[leftIdx] + numbers[rightIdx];

  if (min > Math.abs(sum)) {
    min = Math.abs(sum);
    result = [numbers[leftIdx], numbers[rightIdx]];
  }

  if (sum === 0) break;
  else if (sum > 0) rightIdx--;
  else leftIdx++;
}

console.log(result.join(' '));
